src/Tasks/WriteCodeFragment.cs

```diff
@@ -113,7 +113,9 @@ public override bool Execute()
 
                 OutputFile ??= new TaskItem(FileUtilities.GetTemporaryFile(OutputDirectory.ItemSpec, null, extension));
 
-                FileUtilities.EnsureDirectoryExists(Path.GetDirectoryName(OutputFile.ItemSpec));
+                // To get OutputFile's directory use its full path in case specified OutputFile path is a file name but OutputDirectory is not set
+                var fullPath = FileUtilities.PathIsInvalid(OutputFile.ItemSpec) ? OutputFile.ItemSpec : OutputFile.GetMetadata("FullPath");
```

Comment:
Throwing an exception might break something existing on Unix because of the union of invalid characters of path or file name across OSes https://github.com/dotnet/msbuild/blob/b59f07e4312eb6e3e33e59241453606c81992738/src/Shared/FileUtilities.cs#L94-L114, while Unix just has very fewer invalid characters(`\0`, `/`). I prefer to not introduce breaking change.

Reply:
I approve of the plan to not add breaking changes! I'm just a little confused because if PathIsInvalid(OutputFile.ItemSpec), you then use OutputFile.ItemSpec as the input to Path.GetDirectoryName(fullPath), and that throws an ArgumentException on framework and old .NET Core. It was throwing an exception before, so it isn't like you introduced anything new, but it just seemed like more convoluted logic to end up at the same place in that case, so I was a little confused 🙂

Reply:
> Why not use FullPath unconditionally?

For net472, invalid path causes `OutputFile.GetMetadata("FullPath")` to throw System.InvalidOperationException. So I have to use FullPath conditionally.

Reply:
If it's invalid will `EnsureDirectoryExists` or `WriteAllText` throw the same exception?

Reply:
If it's invalid path `EnsureDirectoryExists` will throw IO related exception, which is the same exception thrown from the same place.

Reply:
@GangWang01 instead of adding this logic, wouldn't the better way of fixing the bug be simply allowing an empty string in `EnsureDirectoryExists`?

In https://github.com/dotnet/msbuild/blob/0326fd7c9e131c4c26bac3c0f72a43ef9fd2812c/src/Shared/FileUtilities.cs#L1488

`if (!string.IsNullOrEmpty(directoryPath)` instead of `if (directoryPath != null`.

Reply:
Ah, you're right! I haven't even noticed this. Thank you!